src/main/java/com/google/jenkins/plugins/k8sengine/KubernetesVerifiers.java

```diff
@@ -0,0 +1,167 @@
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.jenkins.plugins.k8sengine;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.jayway.jsonpath.JsonPath;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+
+/**
+ * Verification adapters for verifying ManifestObjects using kubectl and returning and logging the
+ * result. Inner class, @VerificationResult encapsulates the result and the class will keep a
+ * registry of Verifiers per kubernetes api kind and version.
+ */
+public class KubernetesVerifiers {
+  private static final Logger LOGGER = Logger.getLogger(KubernetesEngineBuilder.class.getName());
+  private static Map<String, Verifier> verifiers = new HashMap<String, Verifier>();
+
+  /**
+   * Represents the result object for verification action. The status field represents whether the
+   * verification succeeded. The toString() method returns the relevant portion of the json for the
+   * verification.
+   */
+  public static class VerificationResult {
+    private String log;
+    private boolean status;
+    private Manifests.ManifestObject manifestObject;
+
+    public String getLog() {
+      return log;
+    }
+
+    public boolean isVerified() {
+      return status;
+    }
+
+    public Manifests.ManifestObject getManifestObject() {
+      return manifestObject;
+    }
+
+    public VerificationResult(String log, boolean status, Manifests.ManifestObject object) {
+      this.log = log;
+      this.status = status;
+      this.manifestObject = object;
+    }
+
+    public String toString() {
+
+      return new StringBuilder()
+          .append(
+              (status)
+                  ? Messages.KubernetesEngineBuilder_VerifyingLogSuccess(manifestObject.describe())
+                  : Messages.KubernetesEngineBuilder_VerifyingLogFailure(manifestObject.describe()))
+          .append("\n")
+          .append(log)
+          .toString();
+    }
+  }
+
+  public interface Verifier {
+    /**
+     * @param kubectl A KubectlWrapper object for querying the object type in the cluster.
+     * @return true If the resource was verified, false otherwise.
+     */
+    VerificationResult verify(KubectlWrapper kubectl, Manifests.ManifestObject object);
+  }
+
+  /** Uses the latest status conditions to verify minimum availability */
+  public static class LatestConditionVerfier implements Verifier {
+    // Currently status conditions are LIFO, and this implementation relies on that.
+    // Could also sort in reverse by timestamp
```

Comment:
By expensive I was mostly referring to the time to run the sort, but if it has catastrophic implications on the engineering effort required here that would also be a good reason to keep things the way they are.

Also, what might the outcome look like if we continued using this implementation but the external behavior changed unexpectedly (however unlikely)?

Reply:
I think that's reasonable feedback.  It's not expensive, but it could be similarly problematic as the verification is dependent on the JSON format anyway (and if using the timestamp the timestamp format) so it's not necessarily clear to me at least which would be more stable.